### PR TITLE
Services unit tests

### PR DESCRIPTION
**Pull request by SenthilKumar-86**
Added void, refund and reversal
Added unit tests for all services

### CODE REVIEW

File: src/Message/AbstractRequest.php [hunk 1]
```diff
@@ -229,6 +229,40 @@ protected function hydrateCardHolderData(DOMDocument $xml)
...
+     * Get the card data.
+     *
+     * Because the heartland gateway uses a common format for passing
+     * card data to the API, this function can be called to get the
+     * data from the associated card object in the format that the
+     * API requires.
+     *
+     * @return array
+     */
+    protected function getCardData() {
```
Comment (slogsdon) [inline src/Message/AbstractRequest.php:243]: Are we using this? I would figure this is covered by `hydrateCardHolderData` and `hydrateManualEntry` methods.

File: src/Message/AbstractRequest.php [hunk 2]
```diff
@@ -507,6 +560,37 @@ public function hydrateDirectMarketData(DOMDocument $xml)
 
         return $directMktDataElement;
     }
+    
+    
+    
+    private function curlreq($url, $headers, $data, $httpVerb){
```
Comment (slogsdon) [inline src/Message/AbstractRequest.php]: This looks to be from testing. Is it necessary to keep here?

File: src/Message/HpsGatewayResponseValidation.php [hunk 3]
```diff
@@ -50,51 +45,56 @@ public static function getException($responseCode, $responseText, $response)
...
             );
         }
 
         return $e;
     }
+    
+    public static function formResponseException($responseException, $exceptionCode = '')
+    {
+        return $responseException;
+    }
```
Comment (slogsdon) [inline src/Message/HpsGatewayResponseValidation.php]: This isn't creating the exception object nor combining/adding the message and the exception code together.

File: src/Message/HpsGatewayResponseValidation.php [hunk 4]
```diff
@@ -50,51 +45,56 @@ public static function getException($responseCode, $responseText, $response)
         case '0':
             break;
         case '-2':
-            $e = new InvalidResponseException(
+            $e = self::formResponseException(
```
Comment (slogsdon) [inline src/Message/HpsGatewayResponseValidation.php]: Let's change these calls to `static::formResponseException` to allow child classes to override `formResponseException` throw late static binding

Comment (slogsdon) [reply src/Message/HpsGatewayResponseValidation.php]: Same applies to `HpsIssuerResponseValidation`

File: src/Message/HpsGatewayResponseValidation.php (hunk 3 above)
Comment (slogsdon) [reply src/Message/HpsGatewayResponseValidation.php]: Same applies to `HpsIssuerResponseValidation`

**Review by slogsdon — CHANGES_REQUESTED** [review]

File: src/Message/AbstractRequest.php (hunk 1 above)
Comment (SenthilKumar-86) [reply src/Message/AbstractRequest.php:243]: This method is used in unit tests, to verify the  card details

File: src/Message/AbstractRequest.php (hunk 2 above)
Comment (SenthilKumar-86) [reply src/Message/AbstractRequest.php]: method is removed now

File: src/Message/HpsGatewayResponseValidation.php (hunk 4 above)
Comment (SenthilKumar-86) [reply src/Message/HpsGatewayResponseValidation.php]: files are updated

File: src/Message/HpsGatewayResponseValidation.php (hunk 3 above)
Comment (SenthilKumar-86) [reply src/Message/HpsGatewayResponseValidation.php]: Currently we are not throwing exception objects. we are just returning the error messages like other payment gateways. this method can be used for future purpose, if we want to combine the message and code we can update this method.

File: src/Message/AbstractRequest.php (hunk 1 above)
Comment (slogsdon) [reply src/Message/AbstractRequest.php:243]: Hmm. I'm not seeing it. Could you link me to it? Also, if it's only used in the tests, is there any way to move this to a test helper or something similar?

Comment (slogsdon) [reply src/Message/AbstractRequest.php:243]: Hmm. I'm not seeing it in the tests. Are you able to link me to it? Also, if it's only used in the tests, is there any way to move this into a test helper or something?

File: src/Message/HpsGatewayResponseValidation.php (hunk 3 above)
Comment (slogsdon) [reply src/Message/HpsGatewayResponseValidation.php]: Understood, but we're no longer creating the exception objects themselves. We went from:

```php
$e = new InvalidResponseException(
    $responseText, HpsExceptionCodes::INVALID_ORIGINAL_TRANSACTION
);
```

to:

```php
$e = self::formResponseException(
    $responseText, HpsExceptionCodes::INVALID_ORIGINAL_TRANSACTION
);
```

So an `InvalidResponseException` object is not being created in `static::formResponseException`. We're just returning the exception message in `static::formResponseException`.

Comment (SenthilKumar-86) [reply src/Message/HpsGatewayResponseValidation.php]: Code updated to return the exception code. User can get this by $response->getReasonCode method

File: src/Message/AbstractRequest.php (hunk 1 above)
Comment (SenthilKumar-86) [reply src/Message/AbstractRequest.php:243]: this method is used in AbstractRequestTest : testCardData UT to verify the card details set by user reflected in gateway

**Review by slogsdon — APPROVED** [review]

**Pull request merged**